components/ws-manager/pkg/manager/monitor.go

```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
 			log.WithError(err).Error("was unable to update pod's start disposal state - this might cause an incorrect disposal state")
 		}
 
-		if doSnapshot {
+		if pvcFeatureEnabled {
+			pvcName := wso.Pod.Name
+			if !createdVolumeSnapshot {
+				// create snapshot object out of PVC
+				volumeSnapshot := &volumesnapshotv1.VolumeSnapshot{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      pvcVolumeSnapshotName,
+						Namespace: m.manager.Config.Namespace,
+						Labels: map[string]string{
+							"workspaceID": workspaceID,
+						},
+					},
+					Spec: volumesnapshotv1.VolumeSnapshotSpec{
+						Source: volumesnapshotv1.VolumeSnapshotSource{
+							PersistentVolumeClaimName: &pvcName,
+						},
+						VolumeSnapshotClassName: &pvcVolumeSnapshotClassName,
+					},
+				}
+
+				err = m.manager.Clientset.Create(ctx, volumeSnapshot)
+				if err != nil {
+					err = xerrors.Errorf("cannot create volumesnapshot: %v", err)
+					return true, nil, err
+				}
+				createdVolumeSnapshot = true
+			}
+			if createdVolumeSnapshot {
+				backoff := wait.Backoff{
+					Steps:    30,
+					Duration: 100 * time.Millisecond,
+					Factor:   1.5,
+					Jitter:   0.1,
+					Cap:      10 * time.Minute,
+				}
+				err = wait.ExponentialBackoff(backoff, func() (bool, error) {
+					var volumeSnapshot volumesnapshotv1.VolumeSnapshot
+					err := m.manager.Clientset.Get(ctx, types.NamespacedName{Namespace: m.manager.Config.Namespace, Name: pvcVolumeSnapshotName}, &volumeSnapshot)
+					if err != nil {
+						if k8serr.IsNotFound(err) {
+							// volumesnapshot doesn't exist yet, retry again
+							return false, nil
+						}
+						log.WithError(err).WithField("VolumeSnapshot.Name", pvcVolumeSnapshotName).Error("was unable to get volume snapshot")
+						return false, err
+					}
+					if volumeSnapshot.Status != nil {
+						if volumeSnapshot.Status.ReadyToUse != nil && *(volumeSnapshot.Status.ReadyToUse) {
+							pvcVolumeSnapshotContentName = *volumeSnapshot.Status.BoundVolumeSnapshotContentName
```

Comment:
In general, the BoundVolumeSnapshotContentName should not be nil.
But it'd be better check the nil first because there probably the `status.ReadyToUse == true` but `status.BoundVolumeSnapshotContentName == nil`.